containers/sprm-to-anndata/context/main.py

```diff
@@ -0,0 +1,93 @@
+import argparse
+from glob import glob
+from pathlib import Path
+from os import makedirs
+import json
+from ast import literal_eval
+
+import anndata
+import zarr
+import pandas as pd
+import numpy as np
+from shapely.geometry import Polygon
+from ast import literal_eval
+
+CELL_X_ANTIGENS_FILE_SUFFIX = ".ome.tiff-cell_channel_mean.csv"
+CLUSTER_FILE_SUFFIX = ".ome.tiff-cell_cluster.csv"
+POLYGON_FILE_SUFFUX = ".ome.tiff-cell_polygons_spatial.csv"
+NUM_VERTICES = 15
+
+
+def downsample_shape(shape):
+    idx = np.round(np.linspace(0, len(shape[0]) - 1, NUM_VERTICES + 1)).astype(int)
+    shape_downsample = np.asarray(shape[0])[idx]
+    poly_downsample = Polygon(shape_downsample).exterior.coords
+    return list(poly_downsample)
+
+
+def get_centroid(shape):
+    poly = Polygon(np.asarray(shape[0]))
+    # poly.centroid.coords is an object that should be converted to a list, but when we get the 0th
+    # coordinate i.e the centroid, it is a tuple which needs to be a list for further processing.
+    return list(list(poly.centroid.coords)[0])
+
+
+def read_sprm_to_pandas(input_file, converters={}):
+    df = pd.read_csv(input_file, converters=converters).set_index("ID")
+    return df
+
+
+def sprm_to_anndata(img_name, input_dir, output_dir):
+    cell_x_antigen_file = Path(input_dir) / Path(img_name + CELL_X_ANTIGENS_FILE_SUFFIX)
+    df_cell_x_antigen = read_sprm_to_pandas(cell_x_antigen_file)
+
+    cluster_file = Path(input_dir) / Path(img_name + CLUSTER_FILE_SUFFIX)
+    # Pandas reads in as int64 by default which can't be interpreted by Javascript.
+    df_cluster = read_sprm_to_pandas(cluster_file).astype("uint8")
+
+    polygon_file = Path(input_dir) / Path(img_name + POLYGON_FILE_SUFFUX)
+    df_spatial = read_sprm_to_pandas(
+        # It seems like the list of points in the "Shape" column is read in as a string
+        # so it needs be evaluated.
+        polygon_file,
+        converters={"Shape": literal_eval},
```

Comment:
Is it really a python literal? This is fairly safe, but if possible, parsing as JSON would be better.

Reply:
Good call.